evm/vm/forks/sharding/vmc_handler.py

```diff
@@ -1,34 +1,147 @@
+        hash32,
+        rlp.sedes.big_endian_int,
+        rlp.sedes.binary,
+    ])
+    header_values = rlp.decode(header_bytes, sedes=sedes)
+    yield 'header', header_values
+    yield 'is_new_head', is_new_head
+    yield 'score', score
+
+
 class VMC(Contract):
 
-    logger = logging.getLogger("evm.chain.sharding.mainchain_handler.VMC")
+    logger = logging.getLogger("evm.chain.sharding.VMC")
+
+    # For handling logs filtering
+    # Event:
+    #   CollationAdded(indexed uint256 shard, bytes collationHeader, bool isNewHead, uint256 score)
+    COLLATION_ADDED_TOPIC = event_signature_to_log_topic(
+        "CollationAdded(int128,bytes4096,bool,int128)"
+    )
 
-    def __init__(self, *args, default_privkey, **kwargs):
+    def __init__(self, *args, log_handler, default_privkey, **kwargs):
+        self.log_handler = log_handler
         self.default_privkey = default_privkey
         self.default_sender_address = default_privkey.public_key.to_canonical_address()
         self.config = get_sharding_config()
+        # shard_id -> list
+        # older <---------------> newer
+        self.new_logs = defaultdict(list)
+        # shard_id -> list
+        # newer <---------------> older
+        self.unchecked_logs = defaultdict(list)
+        # shard_id -> score
+        self.current_score = defaultdict(lambda: None)
+
         super().__init__(*args, **kwargs)
 
+    @to_tuple
+    def _get_new_logs(self, shard_id):
+        shard_id_topic_hex = encode_hex(shard_id.to_bytes(32, byteorder='big'))
+        new_logs = self.log_handler.get_new_logs(
+            address=self.address,
+            topics=[
+                encode_hex(self.COLLATION_ADDED_TOPIC),
+                shard_id_topic_hex,
+            ],
+        )
+        for log in new_logs:
+            yield parse_collation_added_data(log['data'])
+
+    def get_next_log(self, shard_id):
+        new_logs = self._get_new_logs(shard_id)
+        self.new_logs[shard_id].extend(new_logs)
+        if len(self.new_logs[shard_id]) == 0:
+            raise NextLogUnavailable("No more next logs")
+        return self.new_logs[shard_id].pop()
+
+    def fetch_candidate_head(self, shard_id):
+        # Try to return a log that has the score that we are checking for,
+        # checking in order of oldest to most recent.
+        unchecked_logs = pipe(
+            self.unchecked_logs[shard_id],
+            enumerate,
+            tuple,
+            reversed,
+            tuple,
+        )
+        current_score = self.current_score[shard_id]
+
+        for idx, logs_entry in unchecked_logs:
+            if logs_entry['score'] == current_score:
+                return self.unchecked_logs[shard_id].pop(idx)
+        # If no further recorded but unchecked logs exist, go to the next
+        # is_new_head = true log
+        while True:
+            # TODO: currently just raise when there is no log anymore
+            self.unchecked_logs[shard_id].append(self.get_next_log(shard_id))
```

Comment:
Related to my comment about the logs potentially not being maintained in the right order, I think this is the offending line.  It seems like the logs should be pushed onto the front of the list rather than apppended to the end.  If this is the case, then you might consider using a `collections.deque` which has fast inserts and pops to both the beginning and end of the list.

Reply:
Sorry about the late reply. I think I need to think for a while of it tomorrow.
Will soon give an explanation and what to revise. Thank you for pointing these out!